feat(miniflare): Implement Assets RPC Proxy Worker behaviour

CarmenPopoviciu opened the pull request:
will need some follow-up fixture changes once https://github.com/cloudflare/workerd/pull/3742 lands, but we will do that in a separate PR

Fixes DEVX-1643

---

<!--
Please don't delete the checkboxes <3
The following selections do not need to be completed if this PR only contains changes to .md files
-->

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- Wrangler E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required
  - [ ] Not required because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <!--e.g. <https://github.com/cloudflare/cloudflare-docs/pull/>...-->
  - [x] Documentation not necessary because: internal change
- Wrangler V3 Backport
  - [ ] TODO (before merge)
  - [ ] PR: <!--e.g. <https://github.com/cloudflare/workers-sdk/pull/>...-->
  - [x] Not necessary because: experimental feature

<!--
Have you read our [Contributing guide](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md)?
In particular, for non-trivial changes, please always engage on the issue or create a discussion or feature request issue first before writing your code.
-->

petebacondarwin commented on .changeset/real-jobs-own.md:
```diff
@@ -0,0 +1,6 @@
+"wrangler": patch
+---
+
+feat: Implement Proxy Worker behaviour we need for Workers+Assets
```
This is describing the implementation rather than the effect it has on the user's experience.

CarmenPopoviciu replied:
true! will reword

CarmenPopoviciu commented on packages/wrangler/e2e/assets-multiworker.test.ts:
```diff
@@ -349,7 +349,9 @@ describe.each(
 
-			failsIf(style === "dev registry")(".fetch() existing asset", async () => {
+			failsIf(
+				style === "dev registry" && !wranglerDev.includes("--x-assets-rpc")
```
this is good! this tells us that the Proxy Worker solved the issue for these use cases

CarmenPopoviciu commented on fixtures/workers-with-assets-and-service-bindings/tests/index.test.ts:
```diff
@@ -55,7 +59,7 @@ describe.each(devCmds)(
 				// this currently incorrectly returns the User Worker response
 				// instead of the Asset Worker response
-				it.fails(
+				failsIf(!args.length)(
```
this is a good thing! this means that our Proxy Worker solved this use case

dario-piotrowicz replied:
should the comment above be updated as well?

adding something like `unless --x-assets-rpc is provided` at the end?

penalosa commented on fixtures/workers-with-assets-and-service-bindings/tests/index.test.ts:
```diff
@@ -241,6 +245,20 @@ describe.each(devCmds)(
 					);
 				});
+
+				it("should support promise pipelining", async ({ expect }) => {
```
Is this not going to fail?

CarmenPopoviciu replied:
not for named entrypoints, since the Proxy Worker is added to the pipeline only for default exports

penalosa commented on fixtures/workers-with-assets-and-service-bindings/workerA/src/workerD.util.ts:
```diff
@@ -9,13 +9,26 @@ export async function getWorkerDResponses(request: Request, env) {
+	const buzzResult = await foo.bar.buzz();
+
+	// test RpcTarget + promise pipelining
+	const beeCounter = await env.NAMED_ENTRYPOINT.newBeeCounter();
```
```suggestion
	const beeCounter = env.NAMED_ENTRYPOINT.newBeeCounter();
```
?

CarmenPopoviciu replied:
thx for catching that! yes indeed

penalosa commented on fixtures/workers-with-assets-and-service-bindings/workerB-with-default-export/src/index.ts:
```diff
@@ -22,6 +22,17 @@ export default {
 		return `Hello busy ${bee}s from worker-b busyBee(bee)`;
 	},
 
+	/*
+	 * Nested functions
+	 */
```
Could we add a test for properties, as well?

penalosa commented on packages/miniflare/src/workers/assets/rpc-proxy.worker.ts:
```diff
@@ -3,12 +3,53 @@ import type RouterWorker from "@cloudflare/workers-shared/asset-worker/src/index
+				 * Otherwise, forward to the USER_WORKER and return its response
+				 */
+				return function (...args: Array<unknown>) {
+					// @ts-ignore to figure out
```
It should be a `Service` or `Fetcher`, I think? I imagine this will need a ts-expect-error though

CarmenPopoviciu replied:
✅

penalosa commented on packages/miniflare/src/workers/assets/rpc-proxy.worker.ts:
```diff
@@ -3,12 +3,53 @@ import type RouterWorker from "@cloudflare/workers-shared/asset-worker/src/index
+				/*
+				 * Otherwise, forward to the USER_WORKER and return its response
+				 */
+				return function (...args: Array<unknown>) {
```
This doesn't work with properties, but I believe that'll be addressed after the workerd bump?

CarmenPopoviciu replied:
indeed. But I need to first see what I get back from the Runtime, once we have access to that workerd fix

penalosa reviewed (approved):
LGTM

dario-piotrowicz commented on fixtures/workers-with-assets-and-service-bindings/tests/index.test.ts:
```diff
@@ -127,8 +131,8 @@ describe.each(devCmds)(
 				// instead of the Asset Worker response
-				it.fails(
-					"should return Asset Worker response for fetch requestsfor routes that serve static content",
+				failsIf(!args.length)(
```
same as above 🙂

dario-piotrowicz replied:
Also (VERY NIT! 🙈) I think something like this would be much more readable/easy to follow without having to check other parts of this file: `failsIf(!args.includes("--x-assets-rpc"))`

CarmenPopoviciu replied:
totally fair. We are removing the flag though this week, and once we do that, all these `failIf`s won't be necessary anymore

dario-piotrowicz replied:
ah nice ok! 😄

dario-piotrowicz commented on packages/miniflare/src/index.ts:
```diff
@@ -1343,19 +1343,41 @@ export class Miniflare {
+					sockets.push({
+						name,
+						address,
+						service: {
+							name: `${RPC_PROXY_SERVICE_NAME}:${workerOpts.core.name}`,
+						},
```
is the service the only thing that changes in the if vs else brackets? if so, wouldn't it be clearer to keep the call `sockets.push` the exact same but only conditionally swap services? something like this:
![Screenshot 2025-03-19 at 15 48 58](https://github.com/user-attachments/assets/c7354bd4-8d86-420f-a983-80d5b1aee55e)
or
![Screenshot 2025-03-19 at 15 51 24](https://github.com/user-attachments/assets/a51c585e-d132-4e36-a1dc-0a6b9cd9eadb)

CarmenPopoviciu replied:
yeah sure, I can abs do that!

CarmenPopoviciu commented:
@dario-piotrowicz thank you so much for reviewing. I addressed all feedback comments [here](https://github.com/cloudflare/workers-sdk/pull/8376/commits/80891a833673203aef912f42f88a4f2a4270e3cd)

dario-piotrowicz reviewed (approved):
Thanks a lot for the updates @CarmenPopoviciu 🫶 

I'm approving although I haven't been fully followed this RPC proxy worker implementation, but on a surface level the code looks good to me, and I trust Samuel's judgment 😄
